### DISCUSSION TITLE
Handling database dependencies

### DISCUSSION THREAD

**Discussion by buger** [DependencyTrack/dependency-track#2649] [Q&A]
Our software depends on ElasticSearch, PostgreSQL and Mongo. 
I can add postgresql dependency as application type, and set CPE to cpe:/a:postgresql:postgresql:12.1, and it can find CVEs. 

But what is the best practicies on how to maintain it in CI/CD flow. I mean I can automatically generated and upload `go.mod` dependencies, and similar. But database ones are manual ones. 

What is the best practicies to maintain such manually added dependencies? If I upload BOM it removes all the dependencies I created manually 🤔  
Does it mean that I need to maintain multiple BOM files, for example have manual BOM with DB deps, and then use some CLI tooling to merge multiple BOMs together? 
Also any way to populate `Latest Version` field? Maybe I can somehow maintain "Local" repository, from where Dependency Track can read data from? 

Thank you!

**Comment by syalioune** [comment]
Hello buger

> What is the best practicies to maintain such manually added dependencies? If I upload BOM it removes all the dependencies I created manually 🤔
Does it mean that I need to maintain multiple BOM files, for example have manual BOM with DB deps, and then use some CLI tooling to merge multiple BOMs together?

You have two alternatives :

- Creating a parent project (`Project A - vX`) and two child projects (`Project A - Software Deps - vX` & `Project A - Database Deps - vX`). It will allow you to host the two kind of dependencies under the same umbrella project and manage them independently (i.e. BOM upload). However be aware that due to features not yet implemented (e.g. #2410), the parent project will only be a container and won't reflect the overall project risk score

- Using a single project (`Project A - vX`) and use [cyclonedx-cli tool to merge the different BOM](https://github.com/CycloneDX/cyclonedx-cli#merge-command) at build time

Personally, I go by the second approach with a so-called platform BOM (database, buses, ...) which is manually filled and hosted at the root of my application source code repository. The application SBOM is generated during CI, merged with the platform BOM and uploaded into Dependency Track.
However the preferred solution also depend on your responsibility model : Is it the same team that manage applicative and database dependencies or is it different teams ?

> Also any way to populate Latest Version field? Maybe I can somehow maintain "Local" repository, from where Dependency Track can read data from?

In your case, I believe it won't be possible. The `Latest Version` field is populated by DT's internal repository metadata analyzer. It uses `purl` identifiers to search the latest version into well-known public (e.g. `https://proxy.golang.org`) or private dependency repository. If the database components could be represented using `purls` and not `cpe`, it would have been possible to host a mock internal repository and declare it in DT but since you are using `CPE` ((which is normal) you can't.

**Reply by buger** [reply]
I see. Makes sense! 

Is there is a way to display this dependencies in "Dependency Graph" section?

Overall I'm thinking in maintaining manual BOM, and writing a script which will periodically and generate "internal" CVEs about EOL, latest version check and etc, based on https://endoflife.date data.
